### ISSUE TITLE
Feedback on MRR ESPA

### ISSUE THREAD

**Issue by vivian-ng** [vivian-ng/Hardware#2]
Opening a new issue thread to track feedback for the MRR ESPA.

MRR ESPA is the rename of MRR ESP 3DP. The files for MRR ESP 3DP have been migrated to KiCad for v0.8 and those files will be released on GitHub soon. Meanwhile, in preparation for a proper v1.0 release, I have renamed the board to try and give it some identity. :smile: 

MRR ESPA v0.9 will be up on GitHub as soon as I can sort it out. File will be on KiCad so that everyone and modify them easily. All symbols and footprints are either from the generic KiCad libraries, or inside a custom library included with the project. You may need to adjust the paths a bit on your system in order to properly find those files. They are in the subdirectory `projlib` with the footprints in `projfp.pretty` and symbols in `projsym.lib`.

The biggest difference between MRR ESPA v0.9 and MRR ESP 3DP v0.8 would be the separation of the VBED rail. It has been isolated from the VIN rail so that they no longer share a common ground line. This means two different PSUs can be used to power VIN and VBED. This is achieved using an optocoupler to drive the bed MOSFET's gate. A voltage divider to half the VBED voltage is the voltage used to drive the MOSFET. I may want to change this to something closer to one-third; but then, it is easily changed by the end user to whatever ratio they want: just solder the corresponding values of resistors instead of the one given in the schematic.

Will upload the files once I can.

**Comment by vivian-ng** [comment]
Migrating to KiCad ended up in quite a few mistakes. C11 and C12 of v0.8 were connected wrongly, resulting in 5V not going to the 3.3V LDO, and 3.3V not going to the CH340C. These can be fixed by soldering through-hole capacitors in place of these SMD capacitors. Also, the collector and emitter of the S9014 for Q5 needs to be swapped. Again, this can be done by using a through-hole NPN transistor.

The 5V autoselect circuit also didn't seem to want to work. So I just soldered diodes to prevent backflow.

After making these changes (capacitors C11 & C12, Q5, and the diodes to VUSB and 5V1), the board works! However, because IO2 is pulled high, the board cannot be flashed solely with USB. I needed to connect IO2 to GND (via the header pins), reset the board to enter flash mode. Once that is done, the board can be flashed. The next step is to test the motors and MOSFETs. Hopefully, I don't burn anything.

**Comment by vivian-ng** [comment]
I am having some issues trying to compile Marlin 2.0 with TMC2130 for X and Y axes.

`In file included from /tmp/arduino_build_501088/sketch/src/module/../inc/MarlinConfig.h:45:0,
                 from /tmp/arduino_build_501088/sketch/src/module/configuration_store.h:26,
                 from /tmp/arduino_build_501088/sketch/src/module/configuration_store.cpp:47:
/tmp/arduino_build_501088/sketch/src/module/configuration_store.cpp: In function 'void say_M906()':
/tmp/arduino_build_501088/sketch/src/module/configuration_store.cpp:1926:52: error: 'port' was not declared in this scope
     void say_M906(PORTARG_SOLO) { SERIAL_ECHOPGM_P(port, "  M906"); }
                                                    ^
/tmp/arduino_build_501088/sketch/src/module/../inc/../core/serial.h:81:69: note: in definition of macro 'SERIAL_PROTOCOLPGM_P'
   #define SERIAL_PROTOCOLPGM_P(p,x)               (serialprintPGM_P(p,PSTR(x)))
                                                                     ^
/tmp/arduino_build_501088/sketch/src/module/configuration_store.cpp:1926:35: note: in expansion of macro 'SERIAL_ECHOPGM_P'
     void say_M906(PORTARG_SOLO) { SERIAL_ECHOPGM_P(port, "  M906"); }`

I am not really sure how to fix this on Marlin, don't even know if Marlin 2.0's TMC2130 support can even run on the ESP32, so I will leave the hardware option available for the time being while someone eventually works out a software solution to it. Meanwhile, the TMC2130 support on this board will continue to be listed as "Untested" so that people know it is something to be implemented software-wise.

**Comment by vivian-ng** [comment]
Update on v0.9.

Soldered everything, and compiled @luc-github Marlin fork which has his ESP3D interface. Prelim tests show that the compile went okay, I could drive the motors to move using the webUI. I attached a heating block with thermistor to simulate the heat bed, and it was able to heat up too, so at least the optocoupler seems to be working. E0 also heated up.

Only thing that seems to be the problem is AUX1. I can't seem to get my MKS TFT32 to work with the board via AUX1. And in the process, I think I fried my RX and TX pins on the ESP32... so while I can see serial messages on the webUI, I can no longer connect to the board with a wired connection (USB or USB<->Serial module) to see serial messages.

Next step is to find time to wire v0.9 to an actual printer to try some test prints, as well as a stress test (dry-print something for 24 hours or more) if the test prints come out okay. Once everything is working, then I will go back to troubleshooting the MKS TFT32 connection (which probably means I need to solder a new board; good thing I ordered 5 pieces).

Meanwhile, v0.9 files will be released soon. It will be accompanied by a sorting of the directory structure, so let me think through the new directory structure first.

**Comment by vivian-ng** [comment]
Update on v0.9.

The files have been uploaded to the repo under the `current` directory.

Theoretically, v0.9 works as long as you are using external MOSFETs for the bed and hotend. During my testing, I discovered that the traces for the bed and hotend MOSFETs were too thin and also the via connecting top and bottom traces were too small. This causes heat buildup, melts the PCB, and gives off smoke. Not a great thing to do. But with external MOSFETs, the current is taken off the board, so the design works as it is.
BTW, if anyone wants to try out v0.9 using external MOSFETs, you can do away with the bed and hotend MOSFETs onboard to save cost. Just use jumper wires to connect IO2 (hotend), IO4 (bed), and GND (can take from AUX1) to their respective external MOSFETs. Of course, you can leave them in and still use them to drive the external MOSFET.

Also tested motor movement using A4988 drivers, which worked fine. I had issues compiling the webui-enabled Marlin fork by @luc-github using TMC2130 drivers, as mentioned above, but when I used the latest bugsfix-2.0.x from the main Marlin repo, I could compile TMC2130 support with no issue using the [TMCStepper library](https://github.com/teemuatlut/TMCStepper). So theoretically, TMC2130 in SPI mode should work too, although I will need to find time to test this (eventually...)

Also theoretically, TMC2130 and TMC2208 drivers should work in standalone mode if the jumpers underneath the stepstick are removed (MS1 and MS2 must both be tied to VIO to enable 1/16 microstepping). I just ordered two each and will test them when they arrive in a few weeks time.

Way ahead:
I am working on the MOSFET trace issue, and will work on the TMC2130/TMC2208 jumpers to allow them to be used for standalone mode (same 3-pin jumpers, with connections to GND and VIO). The next version of this board will not have the current "use jumpers to configure for TMC2130 SPI" feature. SPI pins are already exposed, so future versions of this board will support standalone mode via jumpers, and SPI mode with the "top-facing pins" method used in most other boards.

The next version of this board should be the "production" version. Once it is done and tested to be respectably working fine, I am going to move to add the I2S support. But will need to check with @simon-jouet on what his I2S setup is like.

**Comment by emre-koc** [comment]
Hi @vivian-ng, any updates on simon-jouet's I2S support. Would like to make a board with ESP32 but I guess simon was not able to look further into this. I would like to contribute if I can.

**Comment by simon-jouet** [comment]
Hey @daboi @vivian-ng,

Yeah I've been far too busy with work, to prototype the ESP32 i2s board, I've got the basic schematic of it done quite some time ago, I just need time to tweak it and layout the PCB (which is very very time consuming). 

I2S does work very well for printing, I've used it with the existing R1 prototype board and some breadboard for some time and didn't run into any issues.

**Comment by emre-koc** [comment]
Hey @simon-jouet! Wasn't expecting fast reply, thanks! Great approach on esp32 btw. I read almost all of your posts on github about this. Do you have a repo for the latest pcb so that I may have contribute to that?

**Comment by simon-jouet** [comment]
@daboi it's a bit hit and miss depends if I have to reply or not when I get the email from github :D. Thanks, glad to hear that, I'm still convinced that the ESP32 is the best way to get a very good and cheap controller.  The updated PCB isn't on github yet considering it still needs work, but I can try to push it to a new branch in the weekend so you (and anybody else interested) can have a look/contribute

**Comment by emre-koc** [comment]
I subscribed almost all topics about this so will keep me up-to-date :)  Happy to hear that. Which repo will you be using for branch?

**Comment by vivian-ng** [comment]
@simon-jouet I just managed to finish soldering the latest version of my own board, which hopefully will solve my burning (literally) issues. Trying to find time to test out the board over the next week. Testing with 12V doesn't really help since it is double the current to get the same power compared to using 24V. Also, I managed to figure out the pins for the I2S from the code:
22 DATA_OUT
25 WS_OUT
26 BCK_OUT

Once the current basic board is tested working, I am going to add two 74HC595s to the board (already got the empty space on the board to place those two ICs; other than capacitors on their VCCs, is there anything else that they will need?) to drive the existing 4 stepper drivers. This should free up enough pins for me to experiment with using those pins for UART/SPI_CS with TMC drivers.

**Comment by simon-jouet** [comment]
> I subscribed almost all topics about this so will keep me up-to-date :) Happy to hear that. Which repo will you be using for branch?

my ESPController repo most likely, I can post here once it's done (just need to remember to do it)

@vivian-ng well done hopefully your testing goes well. Yeah the pins for I2S don't really matter too much as they can be remapped so in the current version I used pins that don't have too many functions (but I should double check because it's a bit hard to remember everything). IIRC there is no need for anything else no, what might be good would be to select whether the 74HC595 are powered by 3.3V or 5V so we can get the speed benefits of 5v if needed but 3.3V for instance if it's used as a reference for a signal back in the ESP32 as it's not 5v tolerant.

**Comment by emre-koc** [comment]
@simon-jouet can we upload gcode file to esp32 from web interface? I saw esp3D repo has that webui but can it work with your implementation? also is it possible to plug an sd card port ? I guess you stream gcode from usb port right? Btw thanks for all info, I am hoping to be a tester of your work =)

**Comment by vivian-ng** [comment]
@daboi I am using the ESP3D-enabled Marlin fork that @luc-github is working on. It works on my test board (ESPA v0.9). v0.9 can work if external MOSFETs are used to take the current off the board; v1.0 will have larger tracks and hopefully eliminate the need for external MOSFETs.

**Comment by emre-koc** [comment]
@vivian-ng thanks for the info. That fork has common code with @simon-jouet's one? I am trying to understand which repo contains which part so sorry for asking in detail. Will you branch from @luc-github's work and add I2S option?  I can definitely try your V1.0 board design.

**Comment by vivian-ng** [comment]
@daboi The fork that @luc-github is working on is based on Marlin bugsfix-2.0.x, which already has I2S implemented by @simon-jouet. I had to figure out the I2S pins from the code because he hardcoded it inside the I2S.cpp file. Maybe if he decides to update the code subsequently, he can consider putting the #defines for those pins in a pins file.

**Comment by emre-koc** [comment]
@vivian-ng thanks for the clarification. So I will be watching your and @simon-jouet's repos and will start to test at some point :) Thanks for the work you guys have done.

**Comment by simon-jouet** [comment]
@daboi @vivian-ng here it is https://github.com/simon-jouet/ESP32Controller/tree/r2 I should have most of the components to prototype that one too, just need to find the time and motivation to layout the PCB (and design the footprints and models for some components...)

**Comment by vivian-ng** [comment]
Tested the board. Partially.

Motors with A4988 drivers working. Endstop working. Hotend was able to heat up to 240 degC and maintain it without any significant rise in the temperature of the MOSFET. So far so good...

And then, when I tried to heat the bed... that was when things got really interesting. Tried room temperature to 40 degC, and saw smoke when it started to heat up to 35 degC. Temperature on the heatsink attached to the MOSFET was reaching 120 degC. Shut off, then I tried gradual heating. Slowly when from room temperature to 35, then step up in steps of 2 to 3 degC. Eventually reached 50 degC. When going to 52 degC, there was a bit of smoke which soon stopped when temperature reached target. Then I tried 55 degC. That was when it started to smoke real bad. Like, things were melting, there was sizzling sound, and lots of smoke. Quickly shut off the mains.

![20190417_120548](https://user-images.githubusercontent.com/24537694/56258103-647ca100-6109-11e9-9746-a78960e0b5fd.jpg)

As you can see, there is a black melted part at the heat bed MOSFET (IRLS3034). Problem is, I cannot figure out why.

The heat bed is 180W, running on 12V, so the current draw (theoretically) is 15A. The MOSFET is driven by a voltage divider that drives the gate at 4.8V. Based on specs, the IRLS3034 should have Rds of 2.0 mOhms at 4.5V. So simple calculation says the MOSFET is dissipating 15x15x0.002 which is 0.45W. At temperature rise of 40degC/W, the MOSFET should only be rising like 18degC. I mean, even given other factors, the temperature rise should not be more than 50degC, right? But I was getting a temp rise of 100degC and more. I can't really figure out why.

Anyway, I will be testing it subsequently with an external MOSFET just to see if I can get something to print.

**Comment by vivian-ng** [comment]
Update on the IRLS3034 MOSFET issue. Ordered IRLS3034 from Mouser, and soldered together another board.
![20190426_164731](https://user-images.githubusercontent.com/24537694/57965748-33e78b80-7983-11e9-9f4e-e00081494b02.jpg)
The one on top is the Mouser one, the one below is the Aliexpress one.

And I finally found time to hook up the board and give it a quick test.
And the Mouser one... works.
![20190518_142426](https://user-images.githubusercontent.com/24537694/57965761-57aad180-7983-11e9-83c2-f1bffa77005b.jpg)
I managed to heat up the heat bed to 70 degC without the MOSFET burning. Without a heat sink on the MOSFET, the temperature of around the MOSFET went to around 110 degC. With that blue heat sink, it kept below 85 degC. Still, the area near the 20A fuse was registering temperatures above 100 degC. I am not sure if the rise in temperature was due to the fuse. This was when I was heating a heat bed at 12V, likely drawing around 15A.

Lesson: Some components can be bought from Aliexpress. But for critical parts (high current stuff, or when very precise values are needed), go for a distributor like Mouser and Digikey.

**Comment by vivian-ng** [comment]
There is a new ESP32 module being released, the [ESP32-S2](https://www.espressif.com/en/news/espressif-announces-%E2%80%A8esp32-s2-secure-wi-fi-mcu).
Apparently, it will be cheaper than the current ESP32, and it will come with USB OTG! There are [some features being taken out](https://www.esp32.com/viewtopic.php?f=2&t=10632), though.

Still, having USB may mean there is no longer a need for the CH340, which will help to reduce the footprint a little.

**Comment by luc-github** [comment]
single core will impact wifi usage / printing

**Comment by vivian-ng** [comment]
I changed the 20A fuse to a Bel 20A 0679L series fuse that I got from Mouser, but there is still a significant temperature rise around the fuse area. Reaches about 140degC when the bed is heating up... but at least nothing is melting away. Still, I think that is because I am using the fuse at quite near its rating; the bed is drawing around 15A, and I should probably be using a 30A fuse (or maybe even 40A) instead of a 20A fuse. A 20A fuse will definitely work for lower currents when using a 24V supply (assuming power rating of the bed remains the same; using a 24V supply on a 12V heat bed will result instead in a current draw of around 30A!!!)

So I am going to swap out the 20A fuse for a 30A one in the next iteration. Which, by the way, will have I2S.

**Comment by vivian-ng** [comment]
I believe I have sorted out the board overheating issues. It has to do with the trace widths for the VBED lines.

For the v1.0 board, I actually just soldered two pieces of copper wire from the terminals to the fuse.
![20190610_212413](https://user-images.githubusercontent.com/24537694/59972003-bd612d80-95c1-11e9-8984-8888a6094502.jpg)

For v1.01 board, I changed the layout a bit to allow traces to the fuse from both upper and lower layers. This allowed the board to stay below 100degC when heating the heat bed @12V (around 15A current). I used a thermocouple to measure the fuse temperature, it registered around 85 degC so it seems to be okay.
![20190621_155703](https://user-images.githubusercontent.com/24537694/59972005-cd790d00-95c1-11e9-86b2-d2656b4062bf.jpg)

v1.0 board works well with 24V, tried it on my Ender-3.
![20190611_115608](https://user-images.githubusercontent.com/24537694/59972006-d4a01b00-95c1-11e9-8992-4dff8bbb3899.jpg)

![20190615_130437](https://user-images.githubusercontent.com/24537694/59972009-de298300-95c1-11e9-9530-1167b2143e9b.jpg)

Obviously, not the best in print quality, but I haven't done any tuning. All I did was use the M503 settings from the original Ender-3 board, and just entered them into the v1.0 board. Still, mesh bed leveling data survived and seemed to be working quite well.

**Comment by vivian-ng** [comment]
I have finally managed to find a PCB manufacturer to get a proper quote. I have launched this project on Kickstarter based on the quote to make it available to everyone who wants one but do not have the time/resources to make one themselves. The reward is priced at cost: 76.5% production and shipping, 5% Kickstarter fees, 4.5% Kickstarter payment provider fees, 6% bank fees, 8% taxes.

For those who already have the required components on hand, making your own board by ordering a PCB (from JLCPCB, for example) is the cheaper option. But for those with no such components, I think backing this Kickstarter is the easy (and cheaper) way since the unit price is based on a larger production volume.

Anyway, here is the link:
[MRR ESPA Kickstarter page](https://www.kickstarter.com/projects/maplerainresearch/mrr-espa)

@luc-github Please do not back this project. I will send one board to you if the project is successful.

**Comment by luc-github** [comment]
@vivian-ng I have added link to the kickstarter in my Fork's readme

**Comment by vivian-ng** [comment]
@luc-github Thank you for the support! Hopefully, the funding goal can be reached. I probably should be trying to advertise the Kickstarter project more aggressively but I don't want to oversell this since it is a community effort and my purpose is just to get it to people who are interested, and not to make people interested.

**Comment by luc-github** [comment]
you do not need to oversell, just communicate project exists ^_^- IMHO some communication would be good  as not so many  esp32 boards available there for not electronician guys

**Comment by vivian-ng** [comment]
I have made the [MRR ESPA](https://github.com/maplerainresearch/MRR_ESPA) v1.3 (non-I2S version) available [here](https://www.maplerain.com/en/products/mrr-espa-prelaunch), and the [MRR ESPE](https://github.com/maplerainresearch/MRR_ESPE) v0.5 (I2S version) [here](https://www.maplerain.com/en/products/mrr-espe-prelaunch) for those who wish to try one without having to get all the parts and solder everything by yourself.

I have been using the non-I2S version for quite a while now and it has been running quite smoothly. My printer is located a distance away from my PC, so while I usually control my printer via the PC, I have an unused smart phone near the printer to let me control the printer from nearby too (such as when doing manual mesh bed leveling or loading filament).

As for the I2S version, I think the board is ready, as it is now, in terms of the hardware. Software support still depends on Marlin development, and I did experience random freezes in the past when using I2S, so I cannot say it is the most stable. I also experienced resets when used with a LCD controller, but recent fixes to the Marlin code seemed to have fixed that. Still, I would say the I2S version is still in the experimental stages in terms of firmware. If stable operation is more important, the MRR ESPA (non-I2S version) is the better option.
